### PR TITLE
add sbml infix parsing and printing

### PR DESCRIPTION
**Pull request by lkeegan**
- add parser function `parse_sbml()`
- add print function `sbml()`
- add `sbml_tokenizer.*`, `sbml_parser.*` to symengine/parser/sbml (see #1780)
- use `api.prefix {sbml}` in sbml_parser.yy to avoid nameclashes with main parser
- sbml_parser.yy is based on #1596
- sbml infix format is described [here](http://model.caltech.edu/software/libsbml/5.18.0/docs/formatted/cpp-api/_l3_formula_formatter_8h.html#a3ee99a0498b6249f6602b98be6733b19)
- apologies for the large PR!

### CODE REVIEW

**Comment by rikardn** [comment]
I am excited about this PR!   I don't think it is related, but we need to do something about the fail of the `test_tutorials`. It is always nice to have everything in green before reviewing PRs.

File: symengine/parser/sbml/sbml_parser.cpp
Comment (rikardn) [inline symengine/parser/sbml/sbml_parser.cpp:191]: Should "log" always be log10? It seems as if this is optional in SBML.

File: symengine/parser/sbml/sbml_parser.cpp
Comment (isuruf) [inline symengine/parser/sbml/sbml_parser.cpp:156]: We can reduce the code duplication here using the same trick as `init_sbml_printer_names`. What do you think?

File: symengine/parser/sbml/sbml_parser.h
Comment (isuruf) [inline symengine/parser/sbml/sbml_parser.h:16]: Can you make this inherit from the string parse so that `parse_numeric` and other functions need not be duplicated?

File: symengine/printers/sbml.cpp
Comment (rikardn) [inline symengine/printers/sbml.cpp:17]: First I thought that it should be `arsinh`, but then I saw that it is actually called `arcsinh` or `asinh` in SBML. Would it be better to use the `asinh` alternative here? In the SBML documentation these forms are listed first.

File: symengine/parser/sbml/sbml_parser.cpp
Comment (lkeegan) [reply symengine/parser/sbml/sbml_parser.cpp:191]: You're right, `log10` is the default but it can optionally instead be `ln` or `error`. There are also other options that are not yet supported here (see http://model.caltech.edu/software/libsbml/5.18.0/docs/formatted/cpp-api/_l3_parser_8cpp.html#a77c72a78afff728a6c569790338f2085), as well as user-defined functions and units.
I would address these in a future PR (I'll open an issue with the missing options/features)

File: symengine/printers/sbml.cpp
Comment (lkeegan) [reply symengine/printers/sbml.cpp:17]: libsbml & jsbml (the two main libraries for working with sbml) both use the long form, I guess to be consistent with mathml:

https://github.com/sbmlteam/mathml-tests/blob/master/tests/mathML_test_0038.xml

File: symengine/parser/sbml/sbml_parser.h
Comment (lkeegan) [reply symengine/parser/sbml/sbml_parser.h]: Yes - after struggling with strange compiler errors it turns out this also required a change to the main parser: https://github.com/symengine/symengine/pull/1785/commits/b9d40f898e83b5bb6e030198fae5a863974c16d0

File: symengine/parser/sbml/sbml_parser.cpp
Comment (lkeegan) [reply symengine/parser/sbml/sbml_parser.cpp:188]: I tried this here: https://github.com/symengine/symengine/pull/1785/commits/a159dcd1464c8cc7847065177e6fa64918790c3b 

Is that what you meant? If so, it does reduce duplication, but it might also require more maintenance - e.g. if a new function is added to the main parser that is not supported by sbml, it would also have to be explicitly removed from the sbml parser. Either way is fine by me though.

File: symengine/parser/sbml/sbml_parser.cpp
Comment (lkeegan) [reply symengine/parser/sbml/sbml_parser.cpp]: I've listed the missing features here: #1786

Comment (rikardn) [reply symengine/parser/sbml/sbml_parser.cpp]: Ok, great!

File: symengine/printers/sbml.cpp
Comment (rikardn) [reply symengine/printers/sbml.cpp:17]: Ok, in that case you should keep it.

File: symengine/parser/sbml/sbml_parser.cpp
Comment (isuruf) [inline symengine/parser/sbml/sbml_parser.cpp:105]: I'm fine with adding these into the main parser.

File: symengine/parser/sbml/sbml_parser.cpp
Comment (rikardn) [reply symengine/parser/sbml/sbml_parser.cpp:188]: Having to remember to remove something in one end when something was added in another sounds like an extra maintenance burden. Don't we want to avoid that?

Comment (isuruf) [reply symengine/parser/sbml/sbml_parser.cpp:188]: Yes, looks good to me.

> if a new function is added to the main parser that is not supported by sbml, it would also have to be explicitly removed from the sbml parser

Is it necessary to remove them though? Would it be bad if they were kept?

Comment (lkeegan) [reply symengine/parser/sbml/sbml_parser.cpp:188]: > Yes, looks good to me.
> 
> > if a new function is added to the main parser that is not supported by sbml, it would also have to be explicitly removed from the sbml parser
> 
> Is it necessary to remove them though? Would it be bad if they were kept?

- it could mean invalid sbml infix gets parsed as valid
  - but it would give an error when printing: so that avoids the invalid sbml from being exported somewhere else
- if a function is added to the main parser with the same name as an existing sbml one it would take precedence (since `insert` is a no-op if the key already exists), which could potentially change the meaning of this function in the sbml parser
  - but this would then cause a sbml_parser test to fail

tldr; no, it's probably fine to keep them

Comment (lkeegan) [reply symengine/parser/sbml/sbml_parser.cpp:188]: I guess the question is what should the sbml parser/printer do with math that the main parser+symengine understands, but that is not part of sbml?

1. Raise an error
  - need to define all valid functions -> code duplication
  - or inherit default functions & remove invalid ones -> maintenance burden when defaults change
2. Fallback to the default printer/parser behaviour
  - no code duplication or maintenance issues

Currently we do `1.` when parsing sbml, and `2.` when printing sbml

- As a user I would prefer option `1.` for both printing & parsing: raising an error is much more helpful than silently parsing/producing invalid infix
- But the current setup is also fine, since the user can just re-parse the printed output to check it is valid
- As a maintainer I would probably prefer to implement this by defining all valid functions (i.e. duplication rather than maintenance) but either way is fine by me

What do you think?

Comment (rikardn) [reply symengine/parser/sbml/sbml_parser.cpp:188]: I also think the parser should throw on illegal sbml. Also the printer should throw for things that cannot be supported or isn't currenty supported in sbml. Some cases I guess could be supported with workarounds if needed. For example we have a printer for Fortran code in Pharmpy that converts `sign(x)` to `ABS(X) / X` (if x is known to be positive) or similar.

File: symengine/parser/sbml/sbml_parser.cpp
Comment (lkeegan) [reply symengine/parser/sbml/sbml_parser.cpp]: done here: https://github.com/symengine/symengine/pull/1785/commits/8b80e6661a27603487cb28cc1426b8f6a669c27e

File: symengine/parser/sbml/sbml_parser.cpp
Comment (lkeegan) [reply symengine/parser/sbml/sbml_parser.cpp:177]: @isuruf what are your thoughts on this?

Comment (isuruf) [reply symengine/parser/sbml/sbml_parser.cpp:177]: I'm okay with either.

Comment (lkeegan) [reply symengine/parser/sbml/sbml_parser.cpp:214]: ok, I'll keep the current behaviour for now then (sbml parser throws on invalid sbml, sbml printer falls back to default printer), and leave improving the printer to throw on invalid sbml for a future PR.

I've reverted to defining all sbml parser functions to avoid having to update the sbml parser when new functions are added to the main parser (https://github.com/symengine/symengine/pull/1785/commits/4054db5bafce480dc825ad113c3109223fc49c8e)

**Comment by lkeegan** [comment]
thanks @rikardn and @isuruf for reviewing! I think this is ready to merge if ok with you?

**Comment by rikardn** [comment]
I think the code is looking good. Would it be possible to keep the code coverage above 90% ?

**Comment by lkeegan** [comment]
@rikardn I added a couple more tests (and found #1804 in the process), but I don't think the coverage will get much higher: nearly all the written code is fully covered - the remaining uncovered code is in the source files auto-generated by re2c/bison & appears to be mostly error handling or implementing things that we don't use: https://app.codecov.io/gh/symengine/symengine/compare/master...sbml_parser/diff

**Review by rikardn — APPROVED** [review]

**Pull request merged**
